### ISSUE TITLE
how to cache html generated from JS

### ISSUE THREAD

**Issue by simran-k** [NekR/offline-plugin#217]
I am using ReactJS SSR and my html is generated from a react component 

1. How do I cache this using offline plugin to enable off line access.
 2. How do I cache my images serviced from cdn? By specifying link to all of those images in externals or is there a better way to do it?

This is what my html generating component looks like. Of course other components are embedded in it. What do I exactly need to do to at least render my home page offline completely.

```
export default class Html extends Component {
  static propTypes = {
    assets: PropTypes.object,
    component: PropTypes.node,
    store: PropTypes.object
  };

  render() {
    const {assets, component, store} = this.props;
    const content = component ? ReactDOM.renderToString(component) : '';
    const head = Helmet.rewind();

    return (
      <html lang="en-us">
        <head>
          {head.base.toComponent()}
          {head.title.toComponent()}
          {head.meta.toComponent()}
          {head.link.toComponent()}
          {head.script.toComponent()}

          <link rel="shortcut icon" href="/favicon.ico" />
          <meta name="viewport" content="width=device-width, initial-scale=1" />
          {/* styles (will be present only in production with webpack extract text plugin) */}
          {Object.keys(assets.styles).map((style, key) =>
            <link href={assets.styles[style]} key={key} media="screen, projection"
                  rel="stylesheet" type="text/css" charSet="UTF-8"/>
          )}

          {/* (will be present only in development mode) */}
          {/* outputs a <style/> tag with all bootstrap styles + App.scss + it could be CurrentPage.scss. */}
          {/* can smoothen the initial style flash (flicker) on page load in development mode. */}
          {/* ideally one could also include here the style for the current page (Home.scss, About.scss, etc) */}
          { Object.keys(assets.styles).length === 0 ? <style dangerouslySetInnerHTML={{__html: require('../theme/bootstrap.config.js') + require('../containers/App/App.scss')._style}}/> : null }
        </head>
        <body>
          <div id="content" dangerouslySetInnerHTML={{__html: content}}/>
          <script dangerouslySetInnerHTML={{__html: `window.__data=${serialize(store.getState())};`}} charSet="UTF-8"/>
          <script src={assets.javascript.main} charSet="UTF-8"/>
        </body>
      </html>
    );
  }
}
```

**Comment by GGAlanSmithee** [comment]
@simran-k 

> How do I cache this using offline plugin to enable off line access.

How your HTML is generated has aboslutley nothing to do with offline-plugin / ServiceWorker. It will cache resources that is served to the client, given your configuration. Can you please be a bit more specific with what problem you are experiencing?

> How do I cache my images serviced from cdn? By specifying link to all of those images in externals or is there a better way to do it?

Yes, given that you have `caches: all`, listing them as `externals` should be it. Not sure if `externals` support wildcards / regex / glob, maybe @NekR can clarify?

**Comment by simran66** [comment]
These are the resources that are cached by service worker. My home page is at localhost:3000/ . How do I cache the page and render it offline?
I see homepage is always cached in offline supported web apps.

![cached](https://cloud.githubusercontent.com/assets/4191009/24289630/1922079e-10a8-11e7-8138-da10b04230eb.png)

**Comment by simran66** [comment]
This is what my configuration looks like for offline plugin in webpack:

```
new OfflinePlugin({
      ServiceWorker: {
            events: true
          },
      //externals:['.'],
      excludes: ['.htaccess'],
      caches: {
        main: [':rest:', './src/'],
        additional: [':externals:']
      },
      safeToUseOptionalCaches:true,
      AppCache: {
        caches: ['main', 'additional', 'optional']
      }
    })
```

**Comment by simran66** [comment]
@GGAlanSmithee : Could I get a direction on this please?

**Comment by GGAlanSmithee** [comment]
@simran66 your `caches.main` looks a bit suspicious, are you really serving your public assets from a folder called `src`?

See https://github.com/NekR/offline-plugin/blob/master/docs/caches.md#cache-sections

Your config depends on how you serve your assets, but try

```javascript
caches: {
        main: ['/', ':rest:']
}
```

**Comment by NekR** [comment]
@simran-k @simran66 lol, are you spamming here with 2 accounts?

**Comment by NekR** [comment]
@GGAlanSmithee I just realized that it's the same person and it's 4th issue by hear about the same thing...

**Comment by GGAlanSmithee** [comment]
@NekR Yeah, I thought the same, a bit weird.

**Comment by GGAlanSmithee** [comment]
@simran-k these questions are answered in the documentation, please read it and you should be able to make it work!

**Comment by simran66** [comment]
@NekR : Lol Naaah! Have different accounts open in diff. browsers

**Comment by simran66** [comment]
@GGAlanSmithee : That's what I thought and tried too. With that I get:

WARNING in OfflinePlugin: Cache asset [/] is not found in the output assets,if it's an external asset, put it to the |externals| option to remove this warning

**Comment by NekR** [comment]
@simran66 that warning already says what to do and you can find answers in the docs. Don't be lazy, read it instead of asking ask about everything.

**Comment by NekR** [comment]
@GGAlanSmithee 
> Not sure if externals support wildcards / regex / glob, maybe @NekR can clarify?

Not it doesn't. That's just static list of externals.

**Comment by NekR** [comment]
@GGAlanSmithee do you think `externals` is unfortunate name? I called it so because things listed there are externals to webpack build assets. Not sure if that's best naming though.

**Comment by simran66** [comment]
@NekR : I am REALLY not being lazy here. Trying to find a solution. I have read your entire documentation. May be I am doomed at this and am killing my time possibly here. 

THE ONLY QUESTION I have been tryin to get you to answer is, 

how do I cache the file/component called on my application startup.

I have tried passing 'localhost:3000' and '/' in externals but could not get it to work. If I was using index.html with script and link tags embedded into into, I could just cache it and was good to go but I rather have a react component here which I have posted above and probably I need to cache that one and can't find out a way. 

If that is too difficult for you to answer without having your opinions out, rather DON'T! 
THANKS FOR YOUR RUSH TO CLOSE THE QUESTIONS.
